chain/car_test.go

```diff
@@ -0,0 +1,190 @@
+	alice := mm.Addresses()[0]
+	cb := chain.NewBuilder(t, address.Undef)
+
+	gene := cb.NewGenesis()
+	ts10 := cb.AppendManyOn(9, gene)
+	msgs := []*types.SignedMessage{
+		mm.NewSignedMessage(alice, 1),
+		mm.NewSignedMessage(alice, 2),
+		mm.NewSignedMessage(alice, 3),
+		mm.NewSignedMessage(alice, 4),
+		mm.NewSignedMessage(alice, 5),
+	}
+	rcts := types.EmptyReceipts(5)
+	ts11 := cb.BuildOneOn(ts10, func(b *chain.BlockBuilder) {
+		b.AddMessages(
+			msgs,
+			rcts,
+		)
+	})
+
+	var buf bytes.Buffer
+	carW := bufio.NewWriter(&buf)
+
+	// export the car file to a buffer
+	exportedKey, err := chain.Export(ctx, ts11, cb, cb, carW)
+	assert.NoError(t, err)
+	assert.Equal(t, ts11.Key(), exportedKey)
+	require.NoError(t, carW.Flush())
+
+	mds := ds.NewMapDatastore()
+	bstore := blockstore.NewBlockstore(mds)
+
+	// import the car file from the buffer
+	carR := bufio.NewReader(&buf)
+	importedKey, err := chain.Import(ctx, bstore, carR)
+	assert.NoError(t, err)
+	assert.Equal(t, ts11.Key(), importedKey)
+
+	// walk the blockstore and assert it had all blocks imported
+	validateBlockstoreImport(t, ts11.Key(), gene.Key(), bstore)
+}
+
+func TestChainImportExportMultiTipSetWithMessages(t *testing.T) {
+	tf.UnitTest(t)
+	ctx := context.Background()
+
+	keys := types.MustGenerateKeyInfo(1, 42)
+	mm := types.NewMessageMaker(t, keys)
+	alice := mm.Addresses()[0]
+	cb := chain.NewBuilder(t, address.Undef)
+
+	gene := cb.NewGenesis()
+	ts10 := cb.AppendManyOn(9, gene)
+	msgs := []*types.SignedMessage{
+		mm.NewSignedMessage(alice, 1),
+		mm.NewSignedMessage(alice, 2),
+		mm.NewSignedMessage(alice, 3),
+		mm.NewSignedMessage(alice, 4),
+		mm.NewSignedMessage(alice, 5),
+	}
+	rcts := types.EmptyReceipts(5)
+	ts11 := cb.BuildOneOn(ts10, func(b *chain.BlockBuilder) {
+		b.AddMessages(
+			msgs,
+			rcts,
+		)
+	})
+
+	ts12 := cb.AppendOn(ts11, 3)
+
+	var buf bytes.Buffer
+	carW := bufio.NewWriter(&buf)
+
+	// export the car file to a buffer
+	exportedKey, err := chain.Export(ctx, ts12, cb, cb, carW)
+	assert.NoError(t, err)
+	assert.Equal(t, ts12.Key(), exportedKey)
+	require.NoError(t, carW.Flush())
+
+	mds := ds.NewMapDatastore()
+	bstore := blockstore.NewBlockstore(mds)
+
+	// import the car file from the buffer
+	carR := bufio.NewReader(&buf)
+	importedKey, err := chain.Import(ctx, bstore, carR)
+	assert.NoError(t, err)
+	assert.Equal(t, ts12.Key(), importedKey)
+
+	// walk the blockstore and assert it had all blocks imported
+	validateBlockstoreImport(t, ts12.Key(), gene.Key(), bstore)
+}
+
+func validateBlockstoreImport(t *testing.T, start, stop types.TipSetKey, bstore blockstore.Blockstore) {
+	// walk the blockstore and assert it had all blocks imported
+	cur := start
+	for !cur.Equals(stop) {
+		parents := cid.NewSet()
+		for _, c := range cur.ToSlice() {
+			bsBlk, err := bstore.Get(c)
+			assert.NoError(t, err)
+			blk, err := types.DecodeBlock(bsBlk.RawData())
+			assert.NoError(t, err)
+
+			if !blk.Messages.Equals(types.EmptyMessagesCID) {
```

Comment:
The store should have an object corresponding to EmptyMessagesCID - make it so and remove the special case. Ditto for receipts.

Reply:
Having trouble groking this comment, EmptyMessage CID's aren't persisted to the store for each block if that's what you are getting at?

Reply:
As discussed in person, I expected the CAR (content-addressed archive) not to store duplicate records with the same content, but I might be wrong. We want to store the empty-message-list object exactly once, but then there's no need for any special cases (for a case that isn't really very special).

Reply:
>  I expected the CAR (content-addressed archive) not to store duplicate records with the same content

From my understanding this is unfortunately not the case. The method ([LDWrite](https://github.com/ipfs/go-car/blob/master/util/util.go#L69)) we are using to persist the chain doesn't deduplicate writes for the same content.

I can modify `chain/car.go` to persist the empty message/receipt collections, but that will mean duplicating their content in the car file for each block containing the empty set.

Reply:
Duplicating the blocks is not a good option. Two ideas:
- Wrap the CAR writer with a stateful structure that records all the keys written and de-dups
- Write the empty collections once at the beginning of export, then retain the export-side special case to avoid duplicating it.

The former removes all the special case handling. The latter removes only the import-side special cases.